### PR TITLE
fix(paragraph): Account for Block in Paragraphs line_width, and line_count methods.

### PR DESCRIPTION
**Pull request by airblast-dev**
Fixes: #1233

The `line_width`, and `line_count` methods for `Paragraph` would not take into account the `Block` if one was set. This will now correctly calculate the values including the `Block`'s width/height.

The `Blocks` newly implemented `horizontal_space`, and `vertical_space` can also be used in `Block::inner` since its mostly duplicate logic. Can add it to this PR or a separate one later. This is also why I decided it would be better to return the results as `(top, bottom)`, and `(left, right)` since it would allow for use in more areas.

### CODE REVIEW

File: src/widgets/block.rs [hunk 1]
```diff
@@ -910,6 +910,41 @@ impl Block<'_> {
...
+        if self.borders.contains(Borders::TOP) || self.has_title_at_position(Position::Top) {
+            top = top.saturating_add(1);
+        }
+
+        if self.borders.contains(Borders::BOTTOM) || self.has_title_at_position(Position::Bottom) {
+            bottom = bottom.saturating_add(1);
+        }
+
+        (top, bottom)
+    }
```
[the remark this answers is not in the document]
Comment (airblast-dev) [reply src/widgets/block.rs:947]: FYI there is a typo in the suggestion in line 939 (`self.padding.bottomr` should be `self.padding.bottom`).

File: src/widgets/block.rs [hunk 2]
```diff
@@ -910,6 +910,41 @@ impl Block<'_> {
...
+
+        if self.borders.contains(Borders::LEFT) {
+            left = left.saturating_add(1);
+        }
+
+        if self.borders.contains(Borders::RIGHT) {
+            right = right.saturating_add(1);
+        }
+
+        (left, right)
```
[the remark this answers is not in the document]
Comment (joshka) [reply src/widgets/block.rs:928]: Mostly I prefer concise code like this over longer code as it's less to have to read / understand.
Don't stress too much about const methods - the function calling this isn't const, so this doesn't need to be.

File: src/widgets/block.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (joshka) [reply src/widgets/block.rs:947]: Fixed the typo

**Review by joshka — APPROVED** [review]

**Review by joshka — COMMENTED** [review]
checks are breaking

**Comment by airblast-dev** [comment]
I accidentally merged instead of rebasing and made a mess. There is also a missing semi-colon in one of the suggestions. Will clean up things locally and do a push.

File: src/widgets/block.rs [hunk 3]
```diff
@@ -1062,6 +1097,128 @@ mod tests {
...
+    #[case::right(Block::new(), Borders::RIGHT, (0, 0))]
+    #[case::bottom(Block::new(), Borders::BOTTOM, (0, 1))]
+    #[case::left(Block::new(), Borders::LEFT, (0, 0))]
+    #[case::top_right(Block::new(), Borders::TOP | Borders::RIGHT, (1, 0))]
+    #[case::top_bottom(Block::new(), Borders::TOP | Borders::BOTTOM, (1, 1))]
+    #[case::top_left(Block::new(), Borders::TOP | Borders::LEFT, (1, 0))]
+    #[case::bottom_right(Block::new(), Borders::BOTTOM | Borders::RIGHT, (0, 1))]
+    #[case::bottom_left(Block::new(), Borders::BOTTOM | Borders::LEFT, (0, 1))]
+    #[case::left_right(Block::new(), Borders::LEFT | Borders::RIGHT, (0, 0))]
+    fn vertical_space_takes_into_account_borders(
```
[the remark this answers is not in the document]
Comment (airblast-dev) [reply src/widgets/block.rs]: Removed the repetition and pushed it.

**Comment by airblast-dev** [comment]
Things should be fixed now. Applied the suggestions, and made a few syntax fixes. I think this is good to go.

**Review by joshka — APPROVED** [review]
Looks great!. Thanks for your contribution :)

**Pull request merged**